Update terraform tutorial docs

thedevelopnik opened the pull request:
Closes #38643

As of v15 the Terraform tutorial is broken, because you just get an error message that `impersonation is not allowed.` You have to use tbot to use Terraform, which is [intended behavior](https://github.com/gravitational/teleport/issues/37627).

This updates the tutorial to use the same methods I used in the most recent webinar in my [standalone k8s operator template](https://github.com/asteroid-earth/standalone-operator-template) which is the best way I can figure to do it, but I'm open to feedback!

Since the tutorial now uses Machine ID, to differentiate the Machine ID/Terraform docs a bit, I've added a concrete, prod-ready example of using it in GitHub Actions, while leaving the original content pretty much alone. I did add one link out to a definition that I thought would be helpful.

This is my first teleport PR, and first docs PR, so I'm sure I've done stuff incorrectly 😄 , let me know what to fix that the CI doesn't.

thedevelopnik committed:
update terraform tutorial to use tbot

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik committed:
update terraform machine-id with github actions example

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik commented:
@ptgott would love your review as well.

zmb3 commented on docs/pages/machine-id/access-guides/terraform.mdx:
```diff
@@ -173,6 +173,153 @@ Check your Teleport cluster, ensuring the role has been created:
+      - name: Install Teleport
+        uses: teleport-actions/setup@v1
+        with:
+          version: 15.0.1
```
Is there a variable we can use for this so that it stays up to date.

thedevelopnik replied:
Great idea. It looks like `(=teleport.version=)` would work but if someone who knows better than me could confirm I'd appreciate it.

thedevelopnik replied:
I've put that variable in for now.

thedevelopnik replied:
It works! Checked it in the preview.

zmb3 commented on docs/pages/machine-id/access-guides/terraform.mdx:
```diff
@@ -173,6 +173,153 @@ Check your Teleport cluster, ensuring the role has been created:
+          PLAN: "terraform\n${{ steps.plan.outputs.stdout }}"
+        with:
+          script: |
+            const output = `#### Terraform Format and Style 🖌\`${{ steps.fmt.outcome }}\`
```
This is nice and fancy but it's also noisy and may distract from what you're actually trying to teach people here (how to use Machine ID with Terraform).

Should we simplify this part?

thedevelopnik replied:
I could take out the Terraform side, and put in a link to Hashi's docs that I pulled them from.

thedevelopnik replied:
Removing this whole addition for now at Paul's suggestion. Resolving.

zmb3 commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -83,66 +81,61 @@ To prepare credentials for a local Teleport user:
+  This script will be used later to clean up our bot in between runs and when we're done.
 
-1. Request a signed identity file for the Terraform user by running the following command:
+1. In a new terminal windwo, run the `create_and_run_terraform_bot.sh` script to create the new bot
```
```suggestion
1. In a new terminal window, run the `create_and_run_terraform_bot.sh` script to create the new bot
```

zmb3 commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
````diff
@@ -190,6 +183,15 @@ To apply the configuration:
    $ terraform apply
    ```
 
+1. When you're done, use CNTL-C to stop the Terraform bot. Then run the `remove_terraform_bot.sh` script to remove the bot and clean up the `terraform-identity` folder:
````
```suggestion
1. When you're done, use Ctrl-C to stop the Terraform bot. Then run the `remove_terraform_bot.sh` script to remove the bot and clean up the `terraform-identity` folder:
```

Just to match what already appears in help messages in the product.

thedevelopnik committed:
change Ctrl statement to match style

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

thedevelopnik committed:
fix spelling

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

thedevelopnik committed:
simplify the actions example, and make version a variable

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik committed:
fix indentation in terraform provider

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik committed:
fix bot name in example command

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

ptgott commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -30,15 +30,13 @@ and Machine ID](../../machine-id/deployment/spacelift.mdx).
-certificates that are appropriate for CI/CD workflows instead of using manually-generated
-credentials that have a longer time-to-live (TTL) period. For more information about
-using Machine ID, see the [Machine ID Getting Started Guide](../../machine-id/getting-started.mdx).
+For production, Terraform should be run using Machine ID on a CI/CD platform.
+For more information on using Machine ID and Terraform in a production setting, see the
+[Machine ID Terraform Guide](../../machine-id/access-guides/terraform.mdx).
```
I would move this to "Next steps" so the reader can sprint through the tutorial first before planning work for production.

thedevelopnik replied:
Updated!

ptgott commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -30,15 +30,13 @@ and Machine ID](../../machine-id/deployment/spacelift.mdx).
-authority to manage resources in the cluster.
-You can create a local Teleport user for this purpose or you can use the machine identity agent
-[(Machine ID)](../../machine-id/introduction.mdx) to generate credentials.
+authority to manage resources in the cluster. Terraform is intended to be used with
+[(Machine ID)](../../machine-id/introduction.mdx), so this tutorial will show you how to
+use a local Machine ID bot to try it out.
```
I think this information about the scope of the guide (i.e., that it includes Machine ID instructions) belongs in the introduction of the guide.

thedevelopnik replied:
Updated!

ptgott commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -51,7 +49,7 @@ To prepare credentials for a local Teleport user:
 1. Create a new file called `terraform.yaml` and open it in an editor.
 
-1. Configure settings for a local Teleport user and role by pasting the following content into the `terraform.yaml` file:
+1. Configure the necessary role by pasting the following content into a `terraform_role.yaml` file:
```
I would clarify what entity will have this role and what it uses the role for

thedevelopnik replied:
Done, let me know what you think of the new language.

ptgott commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -83,66 +81,61 @@ To prepare credentials for a local Teleport user:
 
-   The `terraform` user can't sign in to get credentials, so you must have another user
-   **impersonate** the `terraform` account to request a certificate.
+1. Create a new file called `create_and_run_terraform_bot.sh`, open it in an editor, and
```
I think it would make the guide a little simpler to turn the shell scripts into commands that the reader would paste directly into the terminal at the appropriate pointa in the guide, similar to the way we do it in other docs guides. That way, they wouldn't need to manage files and file permissions.

**Edit:** After finishing the guide, I saw the note that you can re-run the two bot scripts to start/stop the bot. If we're keeping the shell scripts, maybe we could add this clarification earlier in the guide?

thedevelopnik replied:
Added context here. Yeah, this is just a [not friendly process](https://gravitational.slack.com/archives/C08HF8E9F/p1707942263917899), unfortunately.

ptgott commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -83,66 +81,61 @@ To prepare credentials for a local Teleport user:
+   tbot start \
+     --destination-dir=./terraform-identity \
+     --token="$TOKEN" \
+     --auth-server=teleport-ent-15.asteroid.earth:443 \
```
I would clarify that the user should edit this flag

zmb3 replied:
I would also use tele.example.com:443 since that's what we've used in other places in this PR.

thedevelopnik replied:
Good catch, that's copy pasta. Thank you.

zmb3 commented on docs/pages/machine-id/access-guides/terraform.mdx:
```diff
@@ -173,6 +173,91 @@ Check your Teleport cluster, ensuring the role has been created:
+
+Let's look at this in a concrete example of running Terraform in CI using GitHub Actions.
+
+First we'll create a token and bot that allow jobs running in Actions in a specific organization,
```
We don't tell the user how to create these resources, we just include the YAML and expect them to know what to do next.

Are we missing a tctl command?

thedevelopnik replied:
Yes, that's a good call out, I will make sure to include that next time I do something like this.

Removing this whole addition for now at Paul's suggestion. Resolving.

zmb3 commented on docs/pages/machine-id/access-guides/terraform.mdx:
```diff
@@ -173,6 +173,91 @@ Check your Teleport cluster, ensuring the role has been created:
+        id: auth
+        uses: teleport-actions/auth@v2
+        with:
+          # Specify the publically accessible address of your Teleport proxy.
```
```suggestion
          # Specify the publicly accessible address of your Teleport proxy.
```

https://www.grammarphobia.com/blog/2015/08/publicly-publically.html

thedevelopnik replied:
Hah, that's copy pasta from our own doc code on the Actions so gotta fix that somewhere else, I'll find it and note to do later.

thedevelopnik replied:
Removing this whole addition for now at Paul's suggestion. Resolving.

zmb3 commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -83,66 +81,61 @@ To prepare credentials for a local Teleport user:
-       impersonate:
-         users: ['terraform']
-         roles: ['terraform']
+   TOKEN=$(tctl bots add terraform-test --roles=terraform --format=json | jq -r '.token_id')
```
Side note: `tctl tokens add` has a `--format=token` option that just outputs the raw token, making `jq` unnecessary here. We should probably do the same for `tctl bots add` at some point.

thedevelopnik replied:
@zmb3 I got this when trying to use `--format=token`

```
ERROR: enum value must be one of text,json, got 'token'
```

thedevelopnik replied:
Oh you're saying `tokens add`, this is `bots add`. I've been through the rigamorole that ended in "you can't create a token and then give it to the bot as a value.

thedevelopnik commented on docs/pages/machine-id/access-guides/terraform.mdx:
```diff
@@ -94,8 +94,8 @@ outputs:
 If operating `tbot` as a background service, restart it. If running `tbot` in
-one-shot mode, it must be executed before you attempt to execute the Terraform
-plan later.
+[one-shot mode](../../includes/machine-id/daemon-or-oneshot.mdx), it must be executed
```
Ah, reading the style guide now, thank you!

thedevelopnik replied:
Updated!

thedevelopnik committed:
describe one-shot in place instead of linking

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik commented on docs/pages/machine-id/access-guides/terraform.mdx:
````diff
@@ -173,6 +173,91 @@ Check your Teleport cluster, ensuring the role has been created:
 $ tctl get role/terraform-test
 ```
 
+## Example: Github Actions
````
Yeah, that makes sense and was kind of in the back of my mind. I think for me the goal is less to get this specific example in and more to make the Machine-ID/Terraform guide more differentiated. What I can do for now is just revert my changes on that doc so we can get the main tutorial in, since that's the one that's broken.

thedevelopnik replied:
I've removed this section. Only left some context on one-shot that I think is actually helpful for the original content.

thedevelopnik committed:
move production tf/mid note to next steps

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik committed:
move scope of guide info to introduction

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik committed:
give more context to Terraform tutorial instructions

thedevelopnik committed:
remove github actions example from machine id page

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik committed:
add note to replace example server address

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

thedevelopnik commented:
Removed the big change to the Machine-Id/Terraform page at Paul's suggestion that it go somewhere else, which was kind of in the back of my mind anyways. I think there's still a way to differentiate that page better, but I don't know what it is yet, and the important thing is updating the broken tutorial.

Implemented the rest of the feedback!

thedevelopnik committed:
update language on changing example text

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

thedevelopnik committed:
remove disallowed bash script words

Signed-off-by: Dave Sudia <david.sudia@goteleport.com>

strideynet commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -7,6 +7,10 @@ videoBanner: YgNHD4SS8dg
 This guide demonstrates how to: Set up the Terraform provider for Teleport on
 Linux and macOS.
 
+Terraform is intended to be used with Machine ID, Teleport's feature for identifying
```
```suggestion
Terraform is intended to be used with Machine ID, Teleport's feature for providing identities to
```

strideynet commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -83,66 +79,60 @@ To prepare credentials for a local Teleport user:
-1. Configure a role that enables your user to impersonate the Terraform user by pasting
-the following content into the `terraform-impersonator.yaml` file:
+   When run, this creates a new bot with the token retrieved from the JSON output. That token
+   is piped into the start command. For security reasons, we don't allow bot join tokens to be
```
> For security reasons, we don't allow bot join tokens to be created in any way other than adding a bot

I'm not sure this is true - if anything it's an awkward edge in the UX experience at the moment that should be fixed in coming months.

thedevelopnik replied:
Ok, I remember seeing/hearing something that it was for security. Maybe that was Ben making it up too hah.

strideynet commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
It feels like this guide is now pretty heavily overlapped with `/machine-id/access-guides/terraform.mdx` now that we're only recommending Machine ID be used to create credentials for the Terraform provider - I think it'd be useful to distinguish or merge these. WDYT @ptgott

strideynet reviewed (commented):
Looks good - thanks for updating this. I think my main concern is just to double check with @ptgott about docs structure here - the modifications to this guide make it a lot more similar to the existing Machine ID x Terraform guide (albeit, it's a little more "getting started') and I wonder if we need to distinguish one as a "Getting Started" guide.

thedevelopnik committed:
change language about machine id and terraform

Co-authored-by: Noah Stride <noah.stride@goteleport.com>
